Index: src/main/java/org/passay/UsernameRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/UsernameRule.java b/src/main/java/org/passay/rule/UsernameRule.java
rename from src/main/java/org/passay/UsernameRule.java
rename to src/main/java/org/passay/rule/UsernameRule.java
--- a/src/main/java/org/passay/UsernameRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/UsernameRule.java	(date 1653740764636)
@@ -1,5 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/ResourceBundleMessageResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/ResourceBundleMessageResolver.java b/src/main/java/org/passay/resolver/ResourceBundleMessageResolver.java
rename from src/main/java/org/passay/ResourceBundleMessageResolver.java
rename to src/main/java/org/passay/resolver/ResourceBundleMessageResolver.java
--- a/src/main/java/org/passay/ResourceBundleMessageResolver.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/resolver/ResourceBundleMessageResolver.java	(date 1653742658410)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.resolver;
+
+import org.passay.resolver.AbstractMessageResolver;
 
 import java.util.MissingResourceException;
 import java.util.ResourceBundle;
@@ -39,7 +41,7 @@
 
 
   @Override
-  protected String getMessage(final String key)
+  public String getMessage(final String key)
   {
     try {
       return resourceBundle.getString(key);
Index: src/main/java/org/passay/PasswordValidator.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PasswordValidator.java b/src/main/java/org/passay/rule/validator/PasswordValidator.java
rename from src/main/java/org/passay/PasswordValidator.java
rename to src/main/java/org/passay/rule/validator/PasswordValidator.java
--- a/src/main/java/org/passay/PasswordValidator.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/validator/PasswordValidator.java	(date 1653740765160)
@@ -1,13 +1,22 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule.validator;
 
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.List;
 import java.util.stream.Collectors;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 import org.passay.entropy.Entropy;
 import org.passay.entropy.RandomPasswordEntropyFactory;
 import org.passay.entropy.ShannonEntropyFactory;
+import org.passay.resolver.MessageResolver;
+import org.passay.resolver.PropertiesMessageResolver;
+import org.passay.rule.AllowedRegexRule;
+import org.passay.rule.CharacterRule;
+import org.passay.rule.IllegalRegexRule;
+import org.passay.rule.Rule;
 
 /**
  * The central component for evaluating multiple password rules against a candidate password.
@@ -124,7 +133,7 @@
    *
    * @param  passwordData  to estimate entropy for
    *
-   * @throws  IllegalArgumentException  for unknown {@link org.passay.PasswordData.Origin} or if the required {@link
+   * @throws  IllegalArgumentException  for unknown {@link PasswordData.Origin} or if the required {@link
    * CharacterRule} instances are unavailable in the {@link #passwordRules} of this validator instance.
    *
    * @see <a href="http://csrc.nist.gov/publications/nistpubs/800-63-1/SP-800-63-1.pdf">PDF Publication</a>
Index: src/main/java/org/passay/Rule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/Rule.java b/src/main/java/org/passay/rule/Rule.java
rename from src/main/java/org/passay/Rule.java
rename to src/main/java/org/passay/rule/Rule.java
--- a/src/main/java/org/passay/Rule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/Rule.java	(date 1653740765032)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 /**
  * Interface for password strength rules.
Index: src/main/java/org/passay/CharacterCharacteristicsRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CharacterCharacteristicsRule.java b/src/main/java/org/passay/rule/CharacterCharacteristicsRule.java
rename from src/main/java/org/passay/CharacterCharacteristicsRule.java
rename to src/main/java/org/passay/rule/CharacterCharacteristicsRule.java
--- a/src/main/java/org/passay/CharacterCharacteristicsRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/CharacterCharacteristicsRule.java	(date 1653740764714)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.ArrayList;
 import java.util.Arrays;
Index: src/main/java/org/passay/data/character/CyrillicCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/data/character/CyrillicCharacterData.java b/src/main/java/org/passay/data/character/CyrillicCharacterData.java
new file mode 100644
--- /dev/null	(date 1653740764876)
+++ b/src/main/java/org/passay/data/character/CyrillicCharacterData.java	(date 1653740764876)
@@ -0,0 +1,54 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.passay.data.character;
+
+/**
+ * Cyrillic character data.
+ *
+ * @author  Middleware Services
+ */
+public enum CyrillicCharacterData implements CharacterData {
+
+  /**
+   * Lower case characters.
+   */
+  LowerCase("INSUFFICIENT_LOWERCASE", "абвгдеёжзийклмнопрстуфхцчшщъыьэюя"),
+
+  /**
+   * Upper case characters.
+   */
+  UpperCase("INSUFFICIENT_UPPERCASE", "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ");
+
+  /**
+   * Error code.
+   */
+  private final String errorCode;
+
+  /**
+   * Characters.
+   */
+  private final String characters;
+
+  /**
+   * Creates a new cyrillic character data.
+   *
+   * @param code          Error code
+   * @param charString    Characters as string
+   */
+  CyrillicCharacterData(final String code, final String charString)
+  {
+    errorCode = code;
+    characters = charString;
+  }
+
+  @Override
+  public String getErrorCode()
+  {
+    return errorCode;
+  }
+
+  @Override
+  public String getCharacters()
+  {
+    return characters;
+  }
+}
Index: src/main/java/org/passay/PasswordUtils.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PasswordUtils.java b/src/main/java/org/passay/util/PasswordUtils.java
rename from src/main/java/org/passay/PasswordUtils.java
rename to src/main/java/org/passay/util/PasswordUtils.java
--- a/src/main/java/org/passay/PasswordUtils.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/util/PasswordUtils.java	(date 1653740764487)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.util;
 
 import java.util.Arrays;
 
Index: src/test/resources/spring-context.xml
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/resources/spring-context.xml b/src/test/resources/spring-context.xml
--- a/src/test/resources/spring-context.xml	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/resources/spring-context.xml	(date 1653740764601)
@@ -18,13 +18,13 @@
     </constructor-arg>
   </bean>
 
-  <bean id="alphaRule" class="org.passay.CharacterRule"
-    c:data="#{T(org.passay.EnglishCharacterData).Alphabetical}"
+  <bean id="alphaRule" class="org.passay.rule.CharacterRule"
+    c:data="#{T(org.passay.data.character.EnglishCharacterData).Alphabetical}"
     c:num="3" />
-  <bean id="digitRule" class="org.passay.CharacterRule"
-    c:data="#{T(org.passay.EnglishCharacterData).Digit}"
+  <bean id="digitRule" class="org.passay.rule.CharacterRule"
+    c:data="#{T(org.passay.data.character.EnglishCharacterData).Digit}"
     c:num="1" />
-  <bean id="charRule" class="org.passay.CharacterCharacteristicsRule"
+  <bean id="charRule" class="org.passay.rule.CharacterCharacteristicsRule"
     p:numberOfCharacteristics="2">
     <property name="rules">
       <list>
@@ -34,27 +34,27 @@
     </property>
   </bean>
 
-  <bean id="dictRule" class="org.passay.DictionarySubstringRule"
+  <bean id="dictRule" class="org.passay.rule.DictionarySubstringRule"
     p:dictionary-ref="dictionary" />
 
-  <bean id="lenRule" class="org.passay.LengthRule"
+  <bean id="lenRule" class="org.passay.rule.LengthRule"
     p:minimumLength="5"/>
 
-  <bean id="qwertySeqRule" class="org.passay.IllegalSequenceRule"
-    c:data="#{T(org.passay.EnglishSequenceData).USQwerty}" />
+  <bean id="qwertySeqRule" class="org.passay.rule.IllegalSequenceRule"
+    c:data="#{T(org.passay.data.sequence.EnglishSequenceData).USQwerty}" />
 
-  <bean id="alphaSeqRule" class="org.passay.IllegalSequenceRule"
-    c:data="#{T(org.passay.EnglishSequenceData).Alphabetical}" />
+  <bean id="alphaSeqRule" class="org.passay.rule.IllegalSequenceRule"
+    c:data="#{T(org.passay.data.sequence.EnglishSequenceData).Alphabetical}" />
 
-  <bean id="numSeqRule" class="org.passay.IllegalSequenceRule"
-    c:data="#{T(org.passay.EnglishSequenceData).Numerical}" />
+  <bean id="numSeqRule" class="org.passay.rule.IllegalSequenceRule"
+    c:data="#{T(org.passay.data.sequence.EnglishSequenceData).Numerical}" />
 
-  <bean id="repeatRegexRule" class="org.passay.RepeatCharacterRegexRule"/>
+  <bean id="repeatRegexRule" class="org.passay.rule.RepeatCharacterRegexRule"/>
 
-  <bean id="usernameRule" class="org.passay.UsernameRule"
+  <bean id="usernameRule" class="org.passay.rule.UsernameRule"
     p:ignoreCase="true"
     p:matchBackwards="true"/>
 
-  <bean id="wsRule" class="org.passay.WhitespaceRule"/>
+  <bean id="wsRule" class="org.passay.rule.WhitespaceRule"/>
  
 </beans>
Index: src/main/java/org/passay/LengthRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/LengthRule.java b/src/main/java/org/passay/rule/LengthRule.java
rename from src/main/java/org/passay/LengthRule.java
rename to src/main/java/org/passay/rule/LengthRule.java
--- a/src/main/java/org/passay/LengthRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/LengthRule.java	(date 1653740764344)
@@ -1,5 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/GermanCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/GermanCharacterData.java b/src/main/java/org/passay/data/character/GermanCharacterData.java
rename from src/main/java/org/passay/GermanCharacterData.java
rename to src/main/java/org/passay/data/character/GermanCharacterData.java
--- a/src/main/java/org/passay/GermanCharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/GermanCharacterData.java	(date 1653740765195)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
 
 /**
  * german character data.
Index: src/main/java/org/passay/DigestHistoryRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/DigestHistoryRule.java b/src/main/java/org/passay/rule/DigestHistoryRule.java
rename from src/main/java/org/passay/DigestHistoryRule.java
rename to src/main/java/org/passay/rule/DigestHistoryRule.java
--- a/src/main/java/org/passay/DigestHistoryRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/DigestHistoryRule.java	(date 1653740764788)
@@ -1,9 +1,10 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import org.cryptacular.bean.HashBean;
+import org.passay.logic.PasswordData;
 
 /**
  * Rule for determining if a password matches one of any previous digested password a user has chosen. If no password
Index: src/main/java/org/passay/data/sequence/CyrillicSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/data/sequence/CyrillicSequenceData.java b/src/main/java/org/passay/data/sequence/CyrillicSequenceData.java
new file mode 100644
--- /dev/null	(date 1653740764941)
+++ b/src/main/java/org/passay/data/sequence/CyrillicSequenceData.java	(date 1653740764941)
@@ -0,0 +1,55 @@
+/* See LICENSE for licensing and NOTICE for copyright. */
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
+
+/**
+ * Cyrillic character sequence.
+ *
+ * @author Middleware Services
+ */
+public enum CyrillicSequenceData implements SequenceData {
+
+  /**
+   * Alphabetical sequence.
+   */
+  Alphabetical(
+    "ILLEGAL_ALPHABETICAL_SEQUENCE",
+    new CharacterSequence[] {
+      new CharacterSequence("абвгдеёжзийклмнопрстуфхцчшщъыьэюя", "АБВГДЕЁЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯ"),
+    });
+
+  /**
+   * Error code.
+   */
+  private final String errorCode;
+
+  /**
+   * Character sequences.
+   */
+  private final CharacterSequence[] sequences;
+
+  /**
+   * Creates new cyrillic sequence data.
+   *
+   * @param code Error code used for message resolution.
+   * @param seqs One or more character sequences.
+   */
+  CyrillicSequenceData(final String code, final CharacterSequence[] seqs)
+  {
+    errorCode = code;
+    sequences = seqs;
+  }
+
+  @Override
+  public String getErrorCode()
+  {
+    return errorCode;
+  }
+
+  @Override
+  public CharacterSequence[] getSequences()
+  {
+    return sequences;
+  }
+}
Index: src/main/java/org/passay/WhitespaceRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/WhitespaceRule.java b/src/main/java/org/passay/rule/WhitespaceRule.java
rename from src/main/java/org/passay/WhitespaceRule.java
rename to src/main/java/org/passay/rule/WhitespaceRule.java
--- a/src/main/java/org/passay/WhitespaceRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/WhitespaceRule.java	(date 1653740764526)
@@ -1,5 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.util.PasswordUtils;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 
 import java.util.Arrays;
 import java.util.LinkedHashMap;
Index: src/main/java/org/passay/RepeatCharactersRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/RepeatCharactersRule.java b/src/main/java/org/passay/rule/RepeatCharactersRule.java
rename from src/main/java/org/passay/RepeatCharactersRule.java
rename to src/main/java/org/passay/rule/RepeatCharactersRule.java
--- a/src/main/java/org/passay/RepeatCharactersRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/RepeatCharactersRule.java	(date 1653740764336)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.ArrayList;
 import java.util.LinkedHashMap;
Index: src/main/java/org/passay/MatchBehavior.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/MatchBehavior.java b/src/main/java/org/passay/logic/MatchBehavior.java
rename from src/main/java/org/passay/MatchBehavior.java
rename to src/main/java/org/passay/logic/MatchBehavior.java
--- a/src/main/java/org/passay/MatchBehavior.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/logic/MatchBehavior.java	(date 1653740765053)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.logic;
 
 import java.util.function.BiFunction;
 
Index: src/main/java/org/passay/PropertiesMessageResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PropertiesMessageResolver.java b/src/main/java/org/passay/resolver/PropertiesMessageResolver.java
rename from src/main/java/org/passay/PropertiesMessageResolver.java
rename to src/main/java/org/passay/resolver/PropertiesMessageResolver.java
--- a/src/main/java/org/passay/PropertiesMessageResolver.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/resolver/PropertiesMessageResolver.java	(date 1653740764389)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.resolver;
 
 import java.io.IOException;
 import java.io.InputStream;
Index: src/main/java/org/passay/EnglishSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/EnglishSequenceData.java b/src/main/java/org/passay/data/sequence/EnglishSequenceData.java
rename from src/main/java/org/passay/EnglishSequenceData.java
rename to src/main/java/org/passay/data/sequence/EnglishSequenceData.java
--- a/src/main/java/org/passay/EnglishSequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/EnglishSequenceData.java	(date 1653740764738)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
  * English character sequences.
Index: src/main/java/org/passay/MessageResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/MessageResolver.java b/src/main/java/org/passay/resolver/MessageResolver.java
rename from src/main/java/org/passay/MessageResolver.java
rename to src/main/java/org/passay/resolver/MessageResolver.java
--- a/src/main/java/org/passay/MessageResolver.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/resolver/MessageResolver.java	(date 1653740764401)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.resolver;
+
+import org.passay.rule.result.RuleResultDetail;
 
 /**
  * Strategy pattern interface for resolving messages from password validation failures described by a {@link
Index: src/main/java/org/passay/DictionarySubstringRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/DictionarySubstringRule.java b/src/main/java/org/passay/rule/DictionarySubstringRule.java
rename from src/main/java/org/passay/DictionarySubstringRule.java
rename to src/main/java/org/passay/rule/DictionarySubstringRule.java
--- a/src/main/java/org/passay/DictionarySubstringRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/DictionarySubstringRule.java	(date 1653740764797)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import org.passay.dictionary.Dictionary;
 
Index: src/main/java/org/passay/PasswordGenerator.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PasswordGenerator.java b/src/main/java/org/passay/generator/PasswordGenerator.java
rename from src/main/java/org/passay/PasswordGenerator.java
rename to src/main/java/org/passay/generator/PasswordGenerator.java
--- a/src/main/java/org/passay/PasswordGenerator.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/generator/PasswordGenerator.java	(date 1653740764593)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.generator;
+
+import org.passay.rule.CharacterRule;
 
 import java.io.IOException;
 import java.nio.Buffer;
Index: src/main/java/org/passay/GermanSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/GermanSequenceData.java b/src/main/java/org/passay/data/sequence/GermanSequenceData.java
rename from src/main/java/org/passay/GermanSequenceData.java
rename to src/main/java/org/passay/data/sequence/GermanSequenceData.java
--- a/src/main/java/org/passay/GermanSequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/GermanSequenceData.java	(date 1653740764698)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
  * german character sequence.
Index: src/main/java/org/passay/AllowedCharacterRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/AllowedCharacterRule.java b/src/main/java/org/passay/rule/AllowedCharacterRule.java
rename from src/main/java/org/passay/AllowedCharacterRule.java
rename to src/main/java/org/passay/rule/AllowedCharacterRule.java
--- a/src/main/java/org/passay/AllowedCharacterRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/AllowedCharacterRule.java	(date 1653740764319)
@@ -1,5 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.util.PasswordUtils;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 
 import java.util.Arrays;
 import java.util.HashSet;
Index: src/main/java/org/passay/LengthComplexityRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/LengthComplexityRule.java b/src/main/java/org/passay/rule/LengthComplexityRule.java
rename from src/main/java/org/passay/LengthComplexityRule.java
rename to src/main/java/org/passay/rule/LengthComplexityRule.java
--- a/src/main/java/org/passay/LengthComplexityRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/LengthComplexityRule.java	(date 1653740764916)
@@ -1,5 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultDetail;
 
 import java.util.Arrays;
 import java.util.Collections;
Index: src/main/java/org/passay/CyrillicSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CyrillicSequenceData.java b/src/main/java/org/passay/data/sequence/ArchaicCyrillicSequenceData.java
rename from src/main/java/org/passay/CyrillicSequenceData.java
rename to src/main/java/org/passay/data/sequence/ArchaicCyrillicSequenceData.java
--- a/src/main/java/org/passay/CyrillicSequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/ArchaicCyrillicSequenceData.java	(date 1653740765013)
@@ -1,12 +1,14 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
- * Cyrillic character sequence.
+ * Archaic cyrillic character sequence.
  *
  * @author Middleware Services
  */
-public enum CyrillicSequenceData implements SequenceData {
+public enum ArchaicCyrillicSequenceData implements SequenceData {
 
   /**
    * Alphabetical sequence.
@@ -33,7 +35,7 @@
    * @param code Error code used for message resolution.
    * @param seqs One or more character sequences.
    */
-  CyrillicSequenceData(final String code, final CharacterSequence[] seqs)
+  ArchaicCyrillicSequenceData(final String code, final CharacterSequence[] seqs)
   {
     errorCode = code;
     sequences = seqs;
Index: src/main/java/org/passay/DigestDictionaryRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/DigestDictionaryRule.java b/src/main/java/org/passay/rule/DigestDictionaryRule.java
rename from src/main/java/org/passay/DigestDictionaryRule.java
rename to src/main/java/org/passay/rule/DigestDictionaryRule.java
--- a/src/main/java/org/passay/DigestDictionaryRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/DigestDictionaryRule.java	(date 1653740764362)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
Index: src/main/java/org/passay/CyrillicCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CyrillicCharacterData.java b/src/main/java/org/passay/data/character/ArchaicCyrillicCharacterData.java
rename from src/main/java/org/passay/CyrillicCharacterData.java
rename to src/main/java/org/passay/data/character/ArchaicCyrillicCharacterData.java
--- a/src/main/java/org/passay/CyrillicCharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/ArchaicCyrillicCharacterData.java	(date 1653740764458)
@@ -1,12 +1,12 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
 
 /**
- * Cyrillic character data.
+ * Archaic cyrillic character data.
  *
  * @author  Middleware Services
  */
-public enum CyrillicCharacterData implements CharacterData {
+public enum ArchaicCyrillicCharacterData implements CharacterData {
 
   /**
    * Lower case characters.
@@ -34,7 +34,7 @@
    * @param code          Error code
    * @param charString    Characters as string
    */
-  CyrillicCharacterData(final String code, final String charString)
+  ArchaicCyrillicCharacterData(final String code, final String charString)
   {
     errorCode = code;
     characters = charString;
Index: src/test/java/org/passay/entropy/ShannonEntropyFactoryTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/entropy/ShannonEntropyFactoryTest.java b/src/test/java/org/passay/entropy/ShannonEntropyFactoryTest.java
--- a/src/test/java/org/passay/entropy/ShannonEntropyFactoryTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/entropy/ShannonEntropyFactoryTest.java	(date 1653740764616)
@@ -2,8 +2,8 @@
 package org.passay.entropy;
 
 import java.util.List;
-import org.passay.PasswordData;
-import org.passay.Rule;
+import org.passay.logic.PasswordData;
+import org.passay.rule.Rule;
 import org.testng.annotations.Test;
 
 /**
Index: src/test/java/org/passay/entropy/RandomPasswordEntropyFactoryTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/entropy/RandomPasswordEntropyFactoryTest.java b/src/test/java/org/passay/entropy/RandomPasswordEntropyFactoryTest.java
--- a/src/test/java/org/passay/entropy/RandomPasswordEntropyFactoryTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/entropy/RandomPasswordEntropyFactoryTest.java	(date 1653740764825)
@@ -2,8 +2,8 @@
 package org.passay.entropy;
 
 import java.util.List;
-import org.passay.PasswordData;
-import org.passay.Rule;
+import org.passay.logic.PasswordData;
+import org.passay.rule.Rule;
 import org.testng.annotations.Test;
 
 /**
Index: src/test/java/org/passay/HistoryRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/HistoryRuleTest.java b/src/test/java/org/passay/HistoryRuleTest.java
--- a/src/test/java/org/passay/HistoryRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/HistoryRuleTest.java	(date 1653740764729)
@@ -3,7 +3,10 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import org.passay.PasswordData.HistoricalReference;
+
+import org.passay.logic.PasswordData;
+import org.passay.logic.PasswordData.HistoricalReference;
+import org.passay.rule.HistoryRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 
Index: src/main/java/org/passay/IllegalCharacterRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/IllegalCharacterRule.java b/src/main/java/org/passay/rule/IllegalCharacterRule.java
rename from src/main/java/org/passay/IllegalCharacterRule.java
rename to src/main/java/org/passay/rule/IllegalCharacterRule.java
--- a/src/main/java/org/passay/IllegalCharacterRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/IllegalCharacterRule.java	(date 1653740764651)
@@ -1,5 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.util.PasswordUtils;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 
 import java.util.Arrays;
 import java.util.HashSet;
Index: src/test/java/org/passay/WhitespaceRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/WhitespaceRuleTest.java b/src/test/java/org/passay/WhitespaceRuleTest.java
--- a/src/test/java/org/passay/WhitespaceRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/WhitespaceRuleTest.java	(date 1653740764984)
@@ -1,6 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.WhitespaceRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 import org.testng.AssertJUnit;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
Index: src/main/java/org/passay/EnglishCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/EnglishCharacterData.java b/src/main/java/org/passay/data/character/EnglishCharacterData.java
rename from src/main/java/org/passay/EnglishCharacterData.java
rename to src/main/java/org/passay/data/character/EnglishCharacterData.java
--- a/src/main/java/org/passay/EnglishCharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/EnglishCharacterData.java	(date 1653740764659)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
 
 /**
  * English language character data.
Index: src/main/java/org/passay/PasswordData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PasswordData.java b/src/main/java/org/passay/logic/PasswordData.java
rename from src/main/java/org/passay/PasswordData.java
rename to src/main/java/org/passay/logic/PasswordData.java
--- a/src/main/java/org/passay/PasswordData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/logic/PasswordData.java	(date 1653740764904)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.logic;
 
 import java.util.ArrayList;
 import java.util.Arrays;
Index: src/test/java/org/passay/PasswordGeneratorTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/PasswordGeneratorTest.java b/src/test/java/org/passay/PasswordGeneratorTest.java
--- a/src/test/java/org/passay/PasswordGeneratorTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/PasswordGeneratorTest.java	(date 1653740764546)
@@ -3,6 +3,12 @@
 
 import java.util.ArrayList;
 import java.util.List;
+
+import org.passay.logic.PasswordData;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.generator.PasswordGenerator;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
 import org.testng.AssertJUnit;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
Index: src/main/java/org/passay/PolishCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PolishCharacterData.java b/src/main/java/org/passay/data/character/PolishCharacterData.java
rename from src/main/java/org/passay/PolishCharacterData.java
rename to src/main/java/org/passay/data/character/PolishCharacterData.java
--- a/src/main/java/org/passay/PolishCharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/PolishCharacterData.java	(date 1653740764842)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
 
 /**
  * Polish language character data.
Index: src/test/java/org/passay/ResourceBundleMessageResolverTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/ResourceBundleMessageResolverTest.java b/src/test/java/org/passay/ResourceBundleMessageResolverTest.java
--- a/src/test/java/org/passay/ResourceBundleMessageResolverTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/ResourceBundleMessageResolverTest.java	(date 1653740765120)
@@ -3,6 +3,8 @@
 
 import java.util.Locale;
 import java.util.ResourceBundle;
+
+import org.passay.resolver.ResourceBundleMessageResolver;
 import org.testng.AssertJUnit;
 import org.testng.annotations.Test;
 
Index: src/main/java/org/passay/RuleResult.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/RuleResult.java b/src/main/java/org/passay/rule/result/RuleResult.java
rename from src/main/java/org/passay/RuleResult.java
rename to src/main/java/org/passay/rule/result/RuleResult.java
--- a/src/main/java/org/passay/RuleResult.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/result/RuleResult.java	(date 1653740765145)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule.result;
 
 import java.util.ArrayList;
 import java.util.Arrays;
Index: src/test/java/org/passay/AbstractRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/AbstractRuleTest.java b/src/test/java/org/passay/AbstractRuleTest.java
--- a/src/test/java/org/passay/AbstractRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/AbstractRuleTest.java	(date 1653740764309)
@@ -3,6 +3,15 @@
 
 import java.util.Properties;
 import java.util.ResourceBundle;
+
+import org.passay.logic.PasswordData;
+import org.passay.resolver.AbstractMessageResolver;
+import org.passay.resolver.MessageResolver;
+import org.passay.resolver.PropertiesMessageResolver;
+import org.passay.resolver.ResourceBundleMessageResolver;
+import org.passay.rule.Rule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultDetail;
 import org.testng.AssertJUnit;
 import org.testng.annotations.Test;
 
Index: src/main/java/org/passay/CharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CharacterData.java b/src/main/java/org/passay/data/character/CharacterData.java
rename from src/main/java/org/passay/CharacterData.java
rename to src/main/java/org/passay/data/character/CharacterData.java
--- a/src/main/java/org/passay/CharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/CharacterData.java	(date 1653740765130)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
+
+import org.passay.rule.CharacterRule;
 
 /**
  * Input data used by {@link CharacterRule}.
Index: src/test/java/org/passay/NumberRangeRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/NumberRangeRuleTest.java b/src/test/java/org/passay/NumberRangeRuleTest.java
--- a/src/test/java/org/passay/NumberRangeRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/NumberRangeRuleTest.java	(date 1653740764467)
@@ -1,6 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.NumberRangeRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/test/java/org/passay/SourceRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/SourceRuleTest.java b/src/test/java/org/passay/SourceRuleTest.java
--- a/src/test/java/org/passay/SourceRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/SourceRuleTest.java	(date 1653740764754)
@@ -3,7 +3,10 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import org.passay.PasswordData.SourceReference;
+
+import org.passay.logic.PasswordData;
+import org.passay.logic.PasswordData.SourceReference;
+import org.passay.rule.SourceRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 
Index: src/test/java/org/passay/CharacterSequenceTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/CharacterSequenceTest.java b/src/test/java/org/passay/CharacterSequenceTest.java
--- a/src/test/java/org/passay/CharacterSequenceTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/CharacterSequenceTest.java	(date 1653740764499)
@@ -1,12 +1,13 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.CharacterSequence;
 import org.testng.Assert;
 import org.testng.annotations.Test;
 
 
 /**
- * Unit test for {@link org.passay.CharacterSequence}.
+ * Unit test for {@link CharacterSequence}.
  *
  * @author  Middleware Services
  */
@@ -30,7 +31,7 @@
 
 
   /**
-   * Tests the {@link org.passay.CharacterSequence#matches(int, char)} method.
+   * Tests the {@link CharacterSequence#matches(int, char)} method.
    */
   @Test
   public void testMatches()
Index: src/test/java/org/passay/CharacterOccurrencesRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/CharacterOccurrencesRuleTest.java b/src/test/java/org/passay/CharacterOccurrencesRuleTest.java
--- a/src/test/java/org/passay/CharacterOccurrencesRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/CharacterOccurrencesRuleTest.java	(date 1653740764885)
@@ -1,6 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
+import org.passay.rule.CharacterOccurrencesRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/test/java/org/passay/AllowedRegexRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/AllowedRegexRuleTest.java b/src/test/java/org/passay/AllowedRegexRuleTest.java
--- a/src/test/java/org/passay/AllowedRegexRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/AllowedRegexRuleTest.java	(date 1653740764379)
@@ -2,6 +2,9 @@
 package org.passay;
 
 import java.util.regex.Pattern;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.AllowedRegexRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/test/java/org/passay/AllowedCharacterRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/AllowedCharacterRuleTest.java b/src/test/java/org/passay/AllowedCharacterRuleTest.java
--- a/src/test/java/org/passay/AllowedCharacterRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/AllowedCharacterRuleTest.java	(date 1653740764327)
@@ -1,6 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.AllowedCharacterRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 import org.testng.AssertJUnit;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
Index: src/test/java/org/passay/PasswordValidatorTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/PasswordValidatorTest.java b/src/test/java/org/passay/PasswordValidatorTest.java
--- a/src/test/java/org/passay/PasswordValidatorTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/PasswordValidatorTest.java	(date 1653740764952)
@@ -7,14 +7,34 @@
 import org.cryptacular.bean.EncodingHashBean;
 import org.cryptacular.spec.CodecSpec;
 import org.cryptacular.spec.DigestSpec;
-import org.passay.PasswordData.HistoricalReference;
-import org.passay.PasswordData.SourceReference;
+import org.passay.logic.PasswordData;
+import org.passay.logic.PasswordData.HistoricalReference;
+import org.passay.logic.PasswordData.SourceReference;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.data.sequence.EnglishSequenceData;
 import org.passay.dictionary.ArrayWordList;
 import org.passay.dictionary.Dictionary;
 import org.passay.dictionary.WordListDictionary;
 import org.passay.dictionary.WordLists;
 import org.passay.dictionary.sort.ArraysSort;
 import org.passay.entropy.RandomPasswordEntropy;
+import org.passay.rule.AllowedCharacterRule;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.rule.DictionaryRule;
+import org.passay.rule.DictionarySubstringRule;
+import org.passay.rule.DigestHistoryRule;
+import org.passay.rule.DigestSourceRule;
+import org.passay.rule.HistoryRule;
+import org.passay.rule.IllegalSequenceRule;
+import org.passay.rule.LengthRule;
+import org.passay.rule.RepeatCharacterRegexRule;
+import org.passay.rule.Rule;
+import org.passay.rule.SourceRule;
+import org.passay.rule.UsernameRule;
+import org.passay.rule.WhitespaceRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.validator.PasswordValidator;
 import org.testng.AssertJUnit;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
Index: src/test/java/org/passay/DigestDictionaryRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/DigestDictionaryRuleTest.java b/src/test/java/org/passay/DigestDictionaryRuleTest.java
--- a/src/test/java/org/passay/DigestDictionaryRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/DigestDictionaryRuleTest.java	(date 1653740764894)
@@ -4,8 +4,10 @@
 import org.cryptacular.bean.EncodingHashBean;
 import org.cryptacular.spec.CodecSpec;
 import org.cryptacular.spec.DigestSpec;
+import org.passay.logic.PasswordData;
 import org.passay.dictionary.Dictionary;
 import org.passay.dictionary.DictionaryBuilder;
+import org.passay.rule.DigestDictionaryRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Parameters;
Index: src/test/java/org/passay/DigestSourceRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/DigestSourceRuleTest.java b/src/test/java/org/passay/DigestSourceRuleTest.java
--- a/src/test/java/org/passay/DigestSourceRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/DigestSourceRuleTest.java	(date 1653740764746)
@@ -6,6 +6,9 @@
 import org.cryptacular.bean.EncodingHashBean;
 import org.cryptacular.spec.CodecSpec;
 import org.cryptacular.spec.DigestSpec;
+import org.passay.logic.PasswordData;
+import org.passay.rule.DigestSourceRule;
+import org.passay.rule.SourceRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 
Index: src/test/java/org/passay/entropy/AbstractEntropyFactoryTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/entropy/AbstractEntropyFactoryTest.java b/src/test/java/org/passay/entropy/AbstractEntropyFactoryTest.java
--- a/src/test/java/org/passay/entropy/AbstractEntropyFactoryTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/entropy/AbstractEntropyFactoryTest.java	(date 1653740764353)
@@ -3,12 +3,12 @@
 
 import java.util.ArrayList;
 import java.util.List;
-import org.passay.AllowedCharacterRule;
-import org.passay.CharacterCharacteristicsRule;
-import org.passay.CharacterRule;
-import org.passay.EnglishCharacterData;
-import org.passay.PasswordData;
-import org.passay.Rule;
+import org.passay.rule.AllowedCharacterRule;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.logic.PasswordData;
+import org.passay.rule.Rule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 
Index: src/test/java/org/passay/spring/SpringMessageResolverTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/spring/SpringMessageResolverTest.java b/src/test/java/org/passay/resolver/SpringMessageResolverTest.java
rename from src/test/java/org/passay/spring/SpringMessageResolverTest.java
rename to src/test/java/org/passay/resolver/SpringMessageResolverTest.java
--- a/src/test/java/org/passay/spring/SpringMessageResolverTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/resolver/SpringMessageResolverTest.java	(date 1653740764565)
@@ -1,11 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay.spring;
+package org.passay.resolver;
 
 import java.util.Locale;
-import org.passay.LengthRule;
-import org.passay.PasswordData;
-import org.passay.PasswordValidator;
-import org.passay.RuleResult;
+import org.passay.rule.LengthRule;
+import org.passay.logic.PasswordData;
+import org.passay.rule.validator.PasswordValidator;
+import org.passay.rule.result.RuleResult;
 import org.springframework.context.i18n.LocaleContextHolder;
 import org.springframework.context.support.StaticMessageSource;
 import org.testng.AssertJUnit;
Index: src/test/java/org/passay/spring/SpringTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/spring/SpringTest.java b/src/test/java/org/passay/resolver/SpringTest.java
rename from src/test/java/org/passay/spring/SpringTest.java
rename to src/test/java/org/passay/resolver/SpringTest.java
--- a/src/test/java/org/passay/spring/SpringTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/resolver/SpringTest.java	(date 1653740764477)
@@ -1,11 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay.spring;
+package org.passay.resolver;
 
 import java.util.ArrayList;
-import org.passay.PasswordData;
-import org.passay.PasswordValidator;
-import org.passay.Rule;
-import org.passay.RuleResult;
+import org.passay.logic.PasswordData;
+import org.passay.rule.validator.PasswordValidator;
+import org.passay.rule.Rule;
+import org.passay.rule.result.RuleResult;
 import org.springframework.context.support.ClassPathXmlApplicationContext;
 import org.testng.AssertJUnit;
 import org.testng.annotations.Test;
Index: src/test/java/org/passay/LengthRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/LengthRuleTest.java b/src/test/java/org/passay/LengthRuleTest.java
--- a/src/test/java/org/passay/LengthRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/LengthRuleTest.java	(date 1653740765071)
@@ -1,6 +1,10 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
+import org.passay.rule.LengthRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 import org.testng.AssertJUnit;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
Index: src/test/java/org/passay/UsernameRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/UsernameRuleTest.java b/src/test/java/org/passay/UsernameRuleTest.java
--- a/src/test/java/org/passay/UsernameRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/UsernameRuleTest.java	(date 1653740764761)
@@ -1,6 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.UsernameRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/test/java/org/passay/IllegalRegexRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/IllegalRegexRuleTest.java b/src/test/java/org/passay/IllegalRegexRuleTest.java
--- a/src/test/java/org/passay/IllegalRegexRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/IllegalRegexRuleTest.java	(date 1653740764722)
@@ -2,6 +2,9 @@
 package org.passay;
 
 import java.util.regex.Pattern;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.IllegalRegexRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/test/java/org/passay/IllegalCharacterRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/IllegalCharacterRuleTest.java b/src/test/java/org/passay/IllegalCharacterRuleTest.java
--- a/src/test/java/org/passay/IllegalCharacterRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/IllegalCharacterRuleTest.java	(date 1653740764833)
@@ -1,6 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.IllegalCharacterRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultMetadata;
 import org.testng.AssertJUnit;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
Index: src/test/java/org/passay/DigestHistoryRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/DigestHistoryRuleTest.java b/src/test/java/org/passay/DigestHistoryRuleTest.java
--- a/src/test/java/org/passay/DigestHistoryRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/DigestHistoryRuleTest.java	(date 1653740764625)
@@ -7,8 +7,11 @@
 import org.cryptacular.bean.EncodingHashBean;
 import org.cryptacular.spec.CodecSpec;
 import org.cryptacular.spec.DigestSpec;
-import org.passay.PasswordData.HistoricalReference;
-import org.passay.PasswordData.Reference;
+import org.passay.logic.PasswordData;
+import org.passay.logic.PasswordData.HistoricalReference;
+import org.passay.logic.PasswordData.Reference;
+import org.passay.rule.DigestHistoryRule;
+import org.passay.rule.HistoryRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 
Index: src/test/java/org/passay/RepeatCharactersRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/RepeatCharactersRuleTest.java b/src/test/java/org/passay/RepeatCharactersRuleTest.java
--- a/src/test/java/org/passay/RepeatCharactersRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/RepeatCharactersRuleTest.java	(date 1653740764807)
@@ -2,6 +2,9 @@
 package org.passay;
 
 import java.util.Arrays;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.RepeatCharactersRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/main/java/org/passay/entropy/ShannonEntropyFactory.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/entropy/ShannonEntropyFactory.java b/src/main/java/org/passay/entropy/ShannonEntropyFactory.java
--- a/src/main/java/org/passay/entropy/ShannonEntropyFactory.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/entropy/ShannonEntropyFactory.java	(date 1653740764972)
@@ -2,13 +2,13 @@
 package org.passay.entropy;
 
 import java.util.List;
-import org.passay.AbstractDictionaryRule;
-import org.passay.CharacterCharacteristicsRule;
-import org.passay.CharacterRule;
-import org.passay.EnglishCharacterData;
-import org.passay.PasswordData;
-import org.passay.PasswordValidator;
-import org.passay.Rule;
+import org.passay.rule.AbstractDictionaryRule;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.logic.PasswordData;
+import org.passay.rule.validator.PasswordValidator;
+import org.passay.rule.Rule;
 
 /**
  * Factory for creating {@link ShannonEntropy} from password rules and password data.
Index: src/test/java/org/passay/IllegalSequenceRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/IllegalSequenceRuleTest.java b/src/test/java/org/passay/IllegalSequenceRuleTest.java
--- a/src/test/java/org/passay/IllegalSequenceRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/IllegalSequenceRuleTest.java	(date 1653743116057)
@@ -1,6 +1,14 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
+import org.passay.data.sequence.CyrillicSequenceData;
+import org.passay.data.sequence.CzechSequenceData;
+import org.passay.data.sequence.EnglishSequenceData;
+import org.passay.data.sequence.GermanSequenceData;
+import org.passay.data.sequence.PolishSequenceData;
+import org.passay.rule.IllegalSequenceRule;
+import org.passay.rule.validator.PasswordValidator;
 import org.testng.annotations.DataProvider;
 
 /**
@@ -326,6 +334,14 @@
             new IllegalSequenceRule(CyrillicSequenceData.Alphabetical, 5, true, false)),
           new PasswordData("ABCx12y34zcабвгд"),
           codes(CyrillicSequenceData.Alphabetical.getErrorCode()),
+        },
+        // archaic cyrillic and english
+        {
+          new PasswordValidator(
+                  new IllegalSequenceRule(EnglishSequenceData.Alphabetical, 5, true, false),
+                  new IllegalSequenceRule(CyrillicSequenceData.Alphabetical, 5, true, false)),
+          new PasswordData("ABCx12y34zcабвгдіѣѲѴ"),
+          codes(CyrillicSequenceData.Alphabetical.getErrorCode()),
         },
       };
   }
Index: src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java b/src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java
--- a/src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/entropy/RandomPasswordEntropyFactory.java	(date 1653740764575)
@@ -4,11 +4,11 @@
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
-import org.passay.AllowedCharacterRule;
-import org.passay.CharacterCharacteristicsRule;
-import org.passay.CharacterRule;
-import org.passay.PasswordData;
-import org.passay.Rule;
+import org.passay.rule.AllowedCharacterRule;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.logic.PasswordData;
+import org.passay.rule.Rule;
 
 /**
  * Factory for creating {@link RandomPasswordEntropy} from password rules and password data.
Index: src/test/java/org/passay/LengthComplexityRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/LengthComplexityRuleTest.java b/src/test/java/org/passay/LengthComplexityRuleTest.java
--- a/src/test/java/org/passay/LengthComplexityRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/LengthComplexityRuleTest.java	(date 1653740765102)
@@ -2,6 +2,19 @@
 package org.passay;
 
 import java.util.ArrayList;
+
+import org.passay.logic.PasswordData;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.data.sequence.EnglishSequenceData;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.rule.IllegalSequenceRule;
+import org.passay.rule.LengthComplexityRule;
+import org.passay.rule.LengthRule;
+import org.passay.rule.RepeatCharacterRegexRule;
+import org.passay.rule.Rule;
+import org.passay.rule.UsernameRule;
+import org.passay.rule.WhitespaceRule;
 import org.testng.AssertJUnit;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
Index: src/main/java/org/passay/spring/SpringMessageResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/spring/SpringMessageResolver.java b/src/main/java/org/passay/resolver/SpringMessageResolver.java
rename from src/main/java/org/passay/spring/SpringMessageResolver.java
rename to src/main/java/org/passay/resolver/SpringMessageResolver.java
--- a/src/main/java/org/passay/spring/SpringMessageResolver.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/resolver/SpringMessageResolver.java	(date 1653740764666)
@@ -1,10 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay.spring;
+package org.passay.resolver;
 
 import java.util.Locale;
-import org.passay.MessageResolver;
-import org.passay.PropertiesMessageResolver;
-import org.passay.RuleResultDetail;
+
+import org.passay.rule.result.RuleResultDetail;
 import org.springframework.context.MessageSource;
 import org.springframework.context.NoSuchMessageException;
 import org.springframework.context.support.MessageSourceAccessor;
Index: src/test/java/org/passay/DictionarySubstringRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/DictionarySubstringRuleTest.java b/src/test/java/org/passay/DictionarySubstringRuleTest.java
--- a/src/test/java/org/passay/DictionarySubstringRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/DictionarySubstringRuleTest.java	(date 1653740764643)
@@ -2,10 +2,13 @@
 package org.passay;
 
 import java.io.FileReader;
+
+import org.passay.logic.PasswordData;
 import org.passay.dictionary.ArrayWordList;
 import org.passay.dictionary.WordListDictionary;
 import org.passay.dictionary.WordLists;
 import org.passay.dictionary.sort.ArraysSort;
+import org.passay.rule.DictionarySubstringRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Parameters;
Index: src/main/java/org/passay/IllegalRegexRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/IllegalRegexRule.java b/src/main/java/org/passay/rule/IllegalRegexRule.java
rename from src/main/java/org/passay/IllegalRegexRule.java
rename to src/main/java/org/passay/rule/IllegalRegexRule.java
--- a/src/main/java/org/passay/IllegalRegexRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/IllegalRegexRule.java	(date 1653740765171)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.HashSet;
 import java.util.LinkedHashMap;
Index: src/test/java/org/passay/RepeatCharacterRegexRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/RepeatCharacterRegexRuleTest.java b/src/test/java/org/passay/RepeatCharacterRegexRuleTest.java
--- a/src/test/java/org/passay/RepeatCharacterRegexRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/RepeatCharacterRegexRuleTest.java	(date 1653740764706)
@@ -1,6 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
+import org.passay.rule.RepeatCharacterRegexRule;
 import org.testng.annotations.DataProvider;
 
 /**
Index: src/main/java/org/passay/IllegalSequenceRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/IllegalSequenceRule.java b/src/main/java/org/passay/rule/IllegalSequenceRule.java
rename from src/main/java/org/passay/IllegalSequenceRule.java
rename to src/main/java/org/passay/rule/IllegalSequenceRule.java
--- a/src/main/java/org/passay/IllegalSequenceRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/IllegalSequenceRule.java	(date 1653740764537)
@@ -1,5 +1,10 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.CharacterSequence;
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
+import org.passay.data.sequence.SequenceData;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/AbstractMessageResolver.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/AbstractMessageResolver.java b/src/main/java/org/passay/resolver/AbstractMessageResolver.java
rename from src/main/java/org/passay/AbstractMessageResolver.java
rename to src/main/java/org/passay/resolver/AbstractMessageResolver.java
--- a/src/main/java/org/passay/AbstractMessageResolver.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/resolver/AbstractMessageResolver.java	(date 1653740764556)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.resolver;
+
+import org.passay.rule.result.RuleResultDetail;
 
 /**
  * Provides a baseline implementation of {@link #resolve(RuleResultDetail)} which uses {@link String#format(String,
Index: src/main/java/org/passay/NumberRangeRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/NumberRangeRule.java b/src/main/java/org/passay/rule/NumberRangeRule.java
rename from src/main/java/org/passay/NumberRangeRule.java
rename to src/main/java/org/passay/rule/NumberRangeRule.java
--- a/src/main/java/org/passay/NumberRangeRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/NumberRangeRule.java	(date 1653740764444)
@@ -1,5 +1,9 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.MatchBehavior;
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/AbstractDictionaryRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/AbstractDictionaryRule.java b/src/main/java/org/passay/rule/AbstractDictionaryRule.java
rename from src/main/java/org/passay/AbstractDictionaryRule.java
rename to src/main/java/org/passay/rule/AbstractDictionaryRule.java
--- a/src/main/java/org/passay/AbstractDictionaryRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/AbstractDictionaryRule.java	(date 1653740765079)
@@ -1,8 +1,11 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 import org.passay.dictionary.Dictionary;
 
 /**
Index: src/main/java/org/passay/RuleResultDetail.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/RuleResultDetail.java b/src/main/java/org/passay/rule/result/RuleResultDetail.java
rename from src/main/java/org/passay/RuleResultDetail.java
rename to src/main/java/org/passay/rule/result/RuleResultDetail.java
--- a/src/main/java/org/passay/RuleResultDetail.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/result/RuleResultDetail.java	(date 1653740764411)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule.result;
 
 import java.util.Arrays;
 import java.util.LinkedHashMap;
Index: src/main/java/org/passay/AllowedRegexRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/AllowedRegexRule.java b/src/main/java/org/passay/rule/AllowedRegexRule.java
rename from src/main/java/org/passay/AllowedRegexRule.java
rename to src/main/java/org/passay/rule/AllowedRegexRule.java
--- a/src/main/java/org/passay/AllowedRegexRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/AllowedRegexRule.java	(date 1653740765087)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/SourceRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/SourceRule.java b/src/main/java/org/passay/rule/SourceRule.java
rename from src/main/java/org/passay/SourceRule.java
rename to src/main/java/org/passay/rule/SourceRule.java
--- a/src/main/java/org/passay/SourceRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/SourceRule.java	(date 1653740764584)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.LinkedHashMap;
 import java.util.List;
Index: src/main/java/org/passay/CharacterOccurrencesRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CharacterOccurrencesRule.java b/src/main/java/org/passay/rule/CharacterOccurrencesRule.java
rename from src/main/java/org/passay/CharacterOccurrencesRule.java
rename to src/main/java/org/passay/rule/CharacterOccurrencesRule.java
--- a/src/main/java/org/passay/CharacterOccurrencesRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/CharacterOccurrencesRule.java	(date 1653740764674)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.Arrays;
 import java.util.LinkedHashMap;
Index: src/main/java/org/passay/PolishSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PolishSequenceData.java b/src/main/java/org/passay/data/sequence/PolishSequenceData.java
rename from src/main/java/org/passay/PolishSequenceData.java
rename to src/main/java/org/passay/data/sequence/PolishSequenceData.java
--- a/src/main/java/org/passay/PolishSequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/PolishSequenceData.java	(date 1653740765111)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
  * Polish character sequences.
Index: src/main/java/org/passay/PasswordCli.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/PasswordCli.java b/src/main/java/org/passay/PasswordCli.java
--- a/src/main/java/org/passay/PasswordCli.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/PasswordCli.java	(date 1653740764961)
@@ -4,8 +4,22 @@
 import java.io.RandomAccessFile;
 import java.util.ArrayList;
 import java.util.List;
+
+import org.passay.logic.PasswordData;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.data.sequence.EnglishSequenceData;
 import org.passay.dictionary.FileWordList;
 import org.passay.dictionary.TernaryTreeDictionary;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.rule.DictionarySubstringRule;
+import org.passay.rule.IllegalSequenceRule;
+import org.passay.rule.LengthRule;
+import org.passay.rule.RepeatCharacterRegexRule;
+import org.passay.rule.Rule;
+import org.passay.rule.UsernameRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.validator.PasswordValidator;
 
 /**
  * Provides a simple command line interface to password validation.
Index: src/main/java/org/passay/RepeatCharacterRegexRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/RepeatCharacterRegexRule.java b/src/main/java/org/passay/rule/RepeatCharacterRegexRule.java
rename from src/main/java/org/passay/RepeatCharacterRegexRule.java
rename to src/main/java/org/passay/rule/RepeatCharacterRegexRule.java
--- a/src/main/java/org/passay/RepeatCharacterRegexRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/RepeatCharacterRegexRule.java	(date 1653740765004)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 /**
  * Rule for determining if a password contains a duplicate ASCII keyboard sequence. See {@link java.util.regex.Pattern}
Index: src/main/java/org/passay/CharacterRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CharacterRule.java b/src/main/java/org/passay/rule/CharacterRule.java
rename from src/main/java/org/passay/CharacterRule.java
rename to src/main/java/org/passay/rule/CharacterRule.java
--- a/src/main/java/org/passay/CharacterRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/CharacterRule.java	(date 1653740764511)
@@ -1,5 +1,12 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.data.character.CharacterData;
+import org.passay.logic.PasswordData;
+import org.passay.util.PasswordUtils;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultDetail;
+import org.passay.rule.result.RuleResultMetadata;
 
 import java.util.LinkedHashMap;
 import java.util.Map;
Index: src/main/java/org/passay/CzechCharacterData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CzechCharacterData.java b/src/main/java/org/passay/data/character/CzechCharacterData.java
rename from src/main/java/org/passay/CzechCharacterData.java
rename to src/main/java/org/passay/data/character/CzechCharacterData.java
--- a/src/main/java/org/passay/CzechCharacterData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/character/CzechCharacterData.java	(date 1653740764815)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.character;
 
 /**
  * Czech character data.
Index: src/main/java/org/passay/RuleResultMetadata.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/RuleResultMetadata.java b/src/main/java/org/passay/rule/result/RuleResultMetadata.java
rename from src/main/java/org/passay/RuleResultMetadata.java
rename to src/main/java/org/passay/rule/result/RuleResultMetadata.java
--- a/src/main/java/org/passay/RuleResultMetadata.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/result/RuleResultMetadata.java	(date 1653740765042)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule.result;
 
 import java.util.Collections;
 import java.util.HashMap;
Index: src/main/java/org/passay/DigestSourceRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/DigestSourceRule.java b/src/main/java/org/passay/rule/DigestSourceRule.java
rename from src/main/java/org/passay/DigestSourceRule.java
rename to src/main/java/org/passay/rule/DigestSourceRule.java
--- a/src/main/java/org/passay/DigestSourceRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/DigestSourceRule.java	(date 1653740764609)
@@ -1,9 +1,10 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import java.nio.charset.Charset;
 import java.nio.charset.StandardCharsets;
 import org.cryptacular.bean.HashBean;
+import org.passay.logic.PasswordData;
 
 /**
  * Rule for determining if a password matches a digested password from a different source. Useful for when separate
Index: src/main/java/org/passay/CharacterSequence.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CharacterSequence.java b/src/main/java/org/passay/logic/CharacterSequence.java
rename from src/main/java/org/passay/CharacterSequence.java
rename to src/main/java/org/passay/logic/CharacterSequence.java
--- a/src/main/java/org/passay/CharacterSequence.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/logic/CharacterSequence.java	(date 1653740765061)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.logic;
 
 /**
  * Models a sequence of characters in one or more forms as strings of equal length where each string represents one form
Index: src/main/java/org/passay/HistoryRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/HistoryRule.java b/src/main/java/org/passay/rule/HistoryRule.java
rename from src/main/java/org/passay/HistoryRule.java
rename to src/main/java/org/passay/rule/HistoryRule.java
--- a/src/main/java/org/passay/HistoryRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/HistoryRule.java	(date 1653740764851)
@@ -1,5 +1,8 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
+
+import org.passay.logic.PasswordData;
+import org.passay.rule.result.RuleResult;
 
 import java.util.LinkedHashMap;
 import java.util.List;
Index: src/main/java/org/passay/CzechSequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/CzechSequenceData.java b/src/main/java/org/passay/data/sequence/CzechSequenceData.java
rename from src/main/java/org/passay/CzechSequenceData.java
rename to src/main/java/org/passay/data/sequence/CzechSequenceData.java
--- a/src/main/java/org/passay/CzechSequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/CzechSequenceData.java	(date 1653740764682)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
  * Czech character sequence.
Index: src/test/java/org/passay/DictionaryRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/DictionaryRuleTest.java b/src/test/java/org/passay/DictionaryRuleTest.java
--- a/src/test/java/org/passay/DictionaryRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/DictionaryRuleTest.java	(date 1653740764372)
@@ -1,8 +1,10 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
 import org.passay.dictionary.Dictionary;
 import org.passay.dictionary.DictionaryBuilder;
+import org.passay.rule.DictionaryRule;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Parameters;
Index: src/test/java/org/passay/CharacterCharacteristicsRuleTest.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/test/java/org/passay/CharacterCharacteristicsRuleTest.java b/src/test/java/org/passay/CharacterCharacteristicsRuleTest.java
--- a/src/test/java/org/passay/CharacterCharacteristicsRuleTest.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/test/java/org/passay/CharacterCharacteristicsRuleTest.java	(date 1653740765094)
@@ -1,6 +1,13 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
 package org.passay;
 
+import org.passay.logic.PasswordData;
+import org.passay.data.character.EnglishCharacterData;
+import org.passay.rule.CharacterCharacteristicsRule;
+import org.passay.rule.CharacterRule;
+import org.passay.rule.result.RuleResult;
+import org.passay.rule.result.RuleResultDetail;
+import org.passay.rule.result.RuleResultMetadata;
 import org.testng.AssertJUnit;
 import org.testng.annotations.BeforeClass;
 import org.testng.annotations.DataProvider;
Index: src/main/java/org/passay/SequenceData.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/SequenceData.java b/src/main/java/org/passay/data/sequence/SequenceData.java
rename from src/main/java/org/passay/SequenceData.java
rename to src/main/java/org/passay/data/sequence/SequenceData.java
--- a/src/main/java/org/passay/SequenceData.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/data/sequence/SequenceData.java	(date 1653740764771)
@@ -1,5 +1,7 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.data.sequence;
+
+import org.passay.logic.CharacterSequence;
 
 /**
  * Container for one or more {@link CharacterSequence}.
Index: src/main/java/org/passay/DictionaryRule.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/main/java/org/passay/DictionaryRule.java b/src/main/java/org/passay/rule/DictionaryRule.java
rename from src/main/java/org/passay/DictionaryRule.java
rename to src/main/java/org/passay/rule/DictionaryRule.java
--- a/src/main/java/org/passay/DictionaryRule.java	(revision 7f0c0606103c7e68333d91faee6a069ab0ffa244)
+++ b/src/main/java/org/passay/rule/DictionaryRule.java	(date 1653740765023)
@@ -1,5 +1,5 @@
 /* See LICENSE for licensing and NOTICE for copyright. */
-package org.passay;
+package org.passay.rule;
 
 import org.passay.dictionary.Dictionary;
 
